### PR TITLE
[dashboard] add `memray` to requirements.txt

### PR DESCRIPTION
**Pull request by brycehuang30**
## Why are these changes needed?

`memray` is used by the memory profiler but it's currently missing in ray requirements. We will see this error if we try to get the memory profiling from a running actor.

https://github.com/ray-project/ray/assets/18074733/09654b90-2cf1-43e4-8aae-fcbfe3996c0b

## Changes

* Added `memray` to ray dashboard's requirements.
* Run `memray detach` with `sudo` since the attach command is run with `sudo`. 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### Testing Strategy

Manually tested the memory profiling page could be loaded with the fix. Note that `memray` requires `gdb` or `lldb` to work and I install them manually in my system. 

https://github.com/ray-project/ray/assets/18074733/c2c1b127-f4c4-4cff-b9cf-3c403d6ddbda

### CODE REVIEW

**Review by alanwguo — APPROVED** [review]

**Review by architkulkarni — COMMENTED** [review]
We try to keep Ray's dependencies small to prevent the installation size from becoming too large. For example, prometheus and grafana aren't bundled with Ray even though the dashboard uses them.

@richardliaw @rkooo567 do you know if there was a specific reason `memray` wasn't bundled with Ray? I'm fine with including it, but want to make sure we hadn't already decided against it for some reason.

**Comment by rkooo567** [comment]
I think the original intention is users can just download it if they want to use it (so not included by default). I think the overhead of adding it to default dependency is small, but we may need api review

**Comment by architkulkarni** [comment]
Agree it's small (1 MB) and it has only a few dependencies, so I think it's okay.

https://github.com/bloomberg/memray/blob/f0a145ea0bf5ef2a8cee2fd488534f7b55994fca/setup.py#L92-L95

```python
    "jinja2 >= 2.9",
    "typing_extensions; python_version < '3.8.0'",
    "rich >= 11.2.0",
    "textual >= 0.41.0",
```

@richardliaw can we get codeowner approval?

**Review by architkulkarni — APPROVED** [review]

**Review by richardliaw — APPROVED** [review]

**Comment by can-anyscale** [comment]
This is breaking some core cpp tests + window builds (https://buildkite.com/ray-project/postmerge/builds/3790#_). I'm reverting.

**Pull request merged**
